src/base/ClpMap.h

```diff
@@ -0,0 +1,276 @@
+uint64_t
+DefaultMemoryUsage(const Value &e)
+{
+    return sizeof(e);
+}
+
+/// An in-memory associative container enforcing three primary caching policies:
+/// * Capacity: The memory used by cached entries has a configurable limit;
+/// * Lifetime: Entries are hidden (and may be deleted) after their TTL expires;
+/// * Priority: Capacity victims are purged in LRU order.
+/// Individual cache entry operations have average constant-time complexity.
+///
+/// Value must meet STL requirements of Erasable and EmplaceConstructible.
+/// Key must come with std::hash<Key> and std::equal_to<Key> instantiations.
+/// Key::length() must return the number of memory bytes in use by the key.
+/// MemoryUsedBy() must return the number of memory bytes in use by the value.
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &) = DefaultMemoryUsage>
+class ClpMap
+{
+public:
+    /// maximum desired entry caching duration (a.k.a. TTL), in seconds
+    using Ttl = int;
+
+    explicit ClpMap(const uint64_t capacity) { setMemLimit(capacity); }
+    ClpMap(uint64_t capacity, Ttl defaultTtl);
+    ~ClpMap() = default;
+
+    // copying disabled because it may be expensive for large maps
+    // moving (implicitly) disabled for simplicity sake
+    ClpMap(const ClpMap &) = delete;
+    ClpMap &operator =(const ClpMap &) = delete;
+
+    /// \return a pointer to a fresh cached value (or nil)
+    /// The underlying value is owned by the map, so the pointer may be
+    /// invalidated by any non-constant method call, including another get().
+    /// Also moves the found entry to the end of the purging queue.
+    const Value *get(const Key &);
+
+    /// Copy the given value into the map (with the given key and TTL)
+    /// \retval true the value was successfully copied into the map
+    /// \retval false caching was rejected (the map remains unchanged)
+    bool add(const Key &, const Value &, Ttl);
+
+    /// Copy the given value into the map (with the given key and default TTL)
+    bool add(const Key &key, const Value &v) { return add(key, v, defaultTtl_); }
+
+    /// Remove the corresponding entry (if any)
+    void del(const Key &);
+
+    /// Reset the memory capacity for this map, purging if needed
+    void setMemLimit(uint64_t newLimit);
+
+    /// The memory capacity for the map
+    uint64_t memLimit() const { return memLimit_; }
+
+    /// The free space of the map
+    uint64_t freeMem() const { return memLimit() - memoryUsed(); }
+
+    /// The current (approximate) memory usage of the map
+    uint64_t memoryUsed() const { return memUsed_; }
+
+    /// The number of currently stored entries, including expired ones
+    size_t entries() const { return entries_.size(); }
+
+private:
+    /// the keeper of cache entry Key, Value, and caching-related entry metadata
+    class Entry
+    {
+    public:
+        Entry(const Key &aKey, const Value &v, const Ttl ttl): key(aKey), value(v), expires(squid_curtime+ttl) {}
```

Comment:
The `expires` sum will overflow when folks use `ttl=infinity`, right?

Reply:
Yes. Good catch Thank you.

Fixed in b55e5e36e (could not use the SafeSum because time_t is signed).

Reply:
I hope to finish working on this within 24 hours.

Reply:
Still working on this, sorry. Should finish within the next 24 hours.

Reply:
Fixed(?) at 2c941cb. Please test.

Reply:
Thank you. Looks good to me, though signed overflows have never been one of my strong points.

I will spend the next day running this code and see if anything shows up before clearing for merge.